[example] Add rendezvous session to nrf apps

gjc13 opened the pull request:
#### Problem
The BLE connection handler in nrf apps are not using the SPAKE secure session. 

 #### Summary of Changes

- Add secure session handler to the nrf app.
- Modify the initialization flow of `RendezvousSession` to reduce the use of heap.

Also fixes https://github.com/project-chip/connectedhomeip/issues/2902

rwalker-apple commented:
rebased

rwalker-apple commented:
@gjc13 builds failing...

rwalker-apple reviewed (changes requested):
please address build issues

erjiaqing commented:
We need to bypass the pairing procedure, or datamodel messages cannot be accpeted.

gjc13 commented:
> We need to bypass the pairing procedure, or datamodel messages cannot be accpeted.

The datamodel message handling part remains the same using the test key. You can manually set the Thread network use the OpenThread cli to by pass the network provisioning.

rwalker-apple commented:
@gjc13 I think this may need a rebase to address the Cirque failures.  I rebased for you, but I think you may have overwritten that.

rwalker-apple commented:
> @gjc13 I think this may need a rebase to address the Cirque failures. I rebased for you, but I think you may have overwritten that.

nevermind, looks like another Cirque issue.  @erjiaqing @yunhanw-google can you guys have a look?

gjc13 commented:
> > @gjc13 I think this may need a rebase to address the Cirque failures. I rebased for you, but I think you may have overwritten that.
> 
> nevermind, looks like another issue. @erjiaqing @yunhanw-google can you guys have a look to verify it's not Cirque just in case?

I've identified the issue and just added a bypass for the secure session.

rwalker-apple commented on examples/common/chip-app-server/RendezvousServer.cpp:
```diff
@@ -0,0 +1,101 @@
+#include "RendezvousServer.h"
```
why isn't this code finding its way into a CHIP library?

gjc13 replied:
Currently wifi and Thread are sending their network credentials in different format. Wifi is `$ssid:$psk` in text format while Thread the network dataset in binary format.

I believe neither will be the final design for network provisioning and the D&S guys are working on the final network provision profile so we may keep those code here until that general network provision profile is implemented.

andy31415 replied:
I would suggest to add a GH issue and a todo here to remind us to merge the formats.

rwalker-apple commented on src/test_driver/linux-cirque/OnOffClusterTest.sh:
```diff
@@ -39,7 +39,7 @@ function build_chip_lighting() {
     set -x
     cd "$chip_light_dir"
-    gn gen out/debug >/dev/null
+    gn gen out/debug --args='bypass_rendezvous=true' >/dev/null
```
this looks like one of the few places RendezvousServer is being exercised by CI.  if it's turned off here, where is it being tested?

gjc13 replied:
The RendezvousServer has never been tested. In the cirque tests we always manually set the Thread network. Moreover the rendezvous code can only work on BLE before and after the PR so it cannot be tested in the simulation CI which now only supports IP network.

andy31415 replied:
Should we add a GH issue to track this? Or is this nothing that can be done (i.e. CI simulation simply cannot do BLE?)

if there is nothing to be done, some documentation would be nice, maybe where the flag is defined.

rwalker-apple replied:
@gjc13 perhaps instead of a compile-time flag, you can do a runtime check.  I.e. if the Thread network has already been set, there is no need for the Rendezvous server to be exercised this way?

gjc13 replied:
In fact we cannot since in the cirque test we still set the Thread network after the server process is launched.

rwalker-apple commented:
> > > @gjc13 I think this may need a rebase to address the Cirque failures. I rebased for you, but I think you may have overwritten that.
> > 
> > 
> > nevermind, looks like another issue. @erjiaqing @yunhanw-google can you guys have a look to verify it's not Cirque just in case?
> 
> I've identified the issue and just added a bypass for the secure session.

awesome, but have you just disabled any testing of the code you've added?

pan-apple commented:
LGTM
CC: @vivien-apple

vivien-apple commented on src/transport/RendezvousSession.cpp:
```diff
@@ -29,10 +29,11 @@ static const char * kSpake2pKeyExchangeSalt        = "SPAKE2P Key Exchange Salt"
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
+    mParams = params;
```
Before the SetupPINCode was enforced by making sure the constructor receive something. Can we add a VerifyOrExit check with a HasSetupPINCode method in the RendezvousParameters ?
Also we can probably check that the value is not 0 since this one means the accessory is already paired.

gjc13 replied:
Oh I don't know pincode 0 means already paired, thanks for pointing out.

andy31415 commented on examples/common/chip-app-server/Server.cpp:
```diff
@@ -195,21 +108,28 @@ void InitServer()
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
     Optional<Transport::PeerAddress> peer(Transport::Type::kUndefined);
+#ifndef BYPASS_RENDEZVOUS
```
Could we place this together with the code below (eventually within {} if the sucessorexit forces parameter declaration)?

it would make it easier to read if code is localized. We can also split it into a separate function.
It looks like pinCode is also only needed for rendezvous.

bukepo commented on examples/common/chip-app-server/RendezvousServer.cpp:
```diff
@@ -0,0 +1,101 @@
+
+void RendezvousServer::OnRendezvousMessageReceived(PacketBuffer * buffer)
+{
+#if CHIP_ENABLE_OPENTHREAD
```
better to implement a simple Thread operational dataset encoder instead.

bukepo commented on examples/common/chip-app-server/RendezvousServer.cpp:
```diff
@@ -0,0 +1,101 @@
+#include "RendezvousServer.h"
```
Missing license header.
